### PR TITLE
adds spire integration sample README

### PR DESCRIPTION
**Pull request by jufantozzi**
**Please provide a description of this PR:**

Adds instructions for Spire quick-start under `samples/` directory

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [X] Docs
- [ ] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: samples/security/spire/README.md
Comment (jufantozzi) [inline samples/security/spire/README.md]: The link `[Envoy's SDS API integration]` currently points to what we have, regarding the official `istio.io` documentation. This should (prior to merging) be updated to the actual link that will be hosted in istio.io.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: Please always use all capital letters for SPIRE when it appears in text like this. Of course, for cases like spire-agent, lowercase is correct.

File: samples/security/spire/README.md
Comment (jufantozzi) [inline samples/security/spire/README.md]: We could deploy this example without the CSI driver and use the easier daemonset/hostpath setup. This would mean less overhead for the quickstart, but we would be kind of  "teaching bad manners", I suppose? That's why I didn't want to use a simpler deployment setup. Let me know what you guys think about this @maxlambrecht @alexandrealvino @sanderson042

**Review by sanderson042 — COMMENTED** [review]
This is a test to see if commenting is working for me.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: Maybe reword this title to be a little more specific. Here's a first guess:

Integrating SPIRE as a certificate authority in Istio through Envoy SDS

File: samples/security/spire/README.md
Comment (sanderson042) [reply samples/security/spire/README.md]: I agree with your approach of showing the most ideal way to perform the integration. I'd suggest adding a sentence or two about the other approach and linking to that section in the istio.io page if possible. But I'll defer to Max and Alex's opinions on this.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: Typo: whitin -> within
I don't think we need to capitalize Certificate Authority or External CA. (Except CA should be capitalized.)
Maybe add a new sentence at the end of the paragraph that explains why someone would want to use this SPIRE integration instead of the native Istio CA.
I'd also like to see a new paragraph that gives an overview of what this demo is demonstrating to the user, something like, "The steps here demonstrate how SPIRE integrates with Istio to provide an X.509 workload identity certificate to the sample sleep service."

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:11]: I wonder if adding a "Prerequisites" section would help the user? Possible things to put in it:
* Linux (not Windows) environment with Bash
* Kubernetes and Istio environments must be installed (for kubectl and istioctl commands)
* Does there need to be an existing Kubernetes environment set up or does the spire.yaml set it up? Or are we depending on some standard Istio demo environment?
* Does the user need a cloud service or does this run locally like minikube?
* What repo to download to get the spire.yaml and profile.yaml files.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: Let's put the $ before each of these commands to show the command prompt. I think that's an Istio doc convention.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: istio -> Istio

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: I followed the link here. It seems like they went with lowercase: sleep.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md]: Please add "the" before "sleep service"

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:74]: Optionally, you could add a section on how to break down the demo environment.

**Review by sanderson042 — COMMENTED** [review]
Hi Juliano 👋 - here's some suggestions for this doc.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:11]: Regarding the first 4 bullets: 

I was originally very tempted to put something like this, but after taking a look at the similar docs under `/samples` there are no appearances of Kubernetes as pre requisite. I believe it goes without saying since istio is natively for k8s. 

> What repo to download to get the spire.yaml and profile.yaml files.

They will live on the same directory of this README.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md]: I did take a look at most of the other similar README's under /samples and they all go without the $. Do you have any reference for this?

File: samples/security/spire/README.md
Comment (sanderson042) [reply samples/security/spire/README.md:11]: You're right, most of other READMEs in the /samples directory don't include detailed prerequisite info. I did see that [httpbin](https://github.com/istio/istio/tree/master/samples/httpbin) and [tcp-echo](https://github.com/istio/istio/tree/master/samples/tcp-echo) at least say "Install Istio by following the istio install instructions." Also, [kind-lb](https://github.com/istio/istio/tree/master/samples/kind-lb) has a short dependency section.  Maybe we could at least say Istio must be installed. It could be the first step if you want to save space instead of having a Prerequisites section.

I agree the repo info is unnecessary because GitHub will show those files at the top of the page because they're in the same directory. 

Regarding the shell type, I was concerned that some other shells might not support `export` but I see at least zsh does.
 
Most of the other bullets I listed don't seem to be covered in the other /samples READMEs. In my world, they would be. 😄 Looks like the other Mithril doc does have a prerequisites section, so that's good.

File: samples/security/spire/README.md
Comment (maxlambrecht) [reply samples/security/spire/README.md]: IMO the phrase `There are currently multiple forms of integrating third-party Certificate Authorities (CA) whitin Istio.` is irrelevant in this context, we are not explaining in this README which are those multiple forms, but specifically how to integrate with SPIRE.

File: samples/security/spire/README.md
Comment (maxlambrecht) [reply samples/security/spire/README.md]: I think the CSI Driver is a MUST, from the point of view of what I believe the istio maintainers prefer for the socket injection into the pods. The k8s workload registrar is a nice to have.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:11]: > Maybe we could at least say Istio must be installed. It could be the first step if you want to save space instead of having a Prerequisites section.

Regarding this concern, what I think is the most important info here is that SPIRE needs to be up and running prior to installing Istio, which is probably something not as usual for the user and may cause confusion. Do you think that this is clear enough at this point?

File: samples/security/spire/README.md
Comment (sanderson042) [reply samples/security/spire/README.md]: To me the $ helps to show that the blocks with multiple commands are multiple commands and not one long command, like your step 2. Also the Bookinfo README does use $ for commands that users type (though not for command syntax). The main [Istio docs](https://istio.io/latest/docs/examples/bookinfo/#start-the-application-services) include the command prompts as a [standard](https://istio.io/latest/docs/releases/contribute/code-blocks/). But true, most of the other READMEs in /samples don't use the $.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:74]: This will be addressed once the config files are ready.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md]: Thanks @sanderson042, applied all the minor formatting advices, and also summarized the demo overview in the next paragraph.

Good point @maxlambrecht , It's arguable that because it's very short (<1 line) and it gives a contextual intro to the example, it could be valuable... but I guess it's more of a left over from when we were over-explaining a lot... Anyways, I'll remove it 😆

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md]: Giving a deeper thought, 2 separate commands in the same block looks weird to me without the dollar sign, and because we'll have output being showed in those blocks, we'll need to have that differentiated. Thanks!

File: samples/security/spire/README.md
Comment (mchurichi) [inline samples/security/spire/README.md]:
```suggestion
Once SPIRE is deployed and integrated with Istio, we will use a modified version of the [sleep](/samples/sleep/README.md) service to validate that its [Identity](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid) did come from SPIRE. 
```

File: samples/security/spire/README.md
Comment (mchurichi) [inline samples/security/spire/README.md]: can't we just do something like this?
```bash
kubectl wait pod --for=condition=ready -n spire -l app=spire-agent
```
(not sure if it works, but worth a try)

File: samples/security/spire/README.md
Comment (mchurichi) [inline samples/security/spire/README.md]:
```suggestion
    $ kubectl apply -f <(istioctl kube-inject -f sleep.yaml)
```
we should be consistent on prompt usage

Comment (jufantozzi) [reply samples/security/spire/README.md]: missed this one, thanks

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md]: nice one, it does work

**Review by alexandrealvino — APPROVED** [review]
Well done 🥇

**Review by maxlambrecht — APPROVED** [review]
LGTM! 🍰

File: samples/security/spire/README.md
Comment (sanderson042) [reply samples/security/spire/README.md:11]: With the current flow of the doc, I don't think a prerequisites section is needed anymore. Users can just quickly read through the steps to understand what's needed.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:51]: Looking at the GitHub Markdown preview, this command is pretty long and, at least on my window, I have to scroll horizontally to read it all. Maybe add a continuation break after the -r so users don't have to scroll? I mean to put a \ after -r and put the rest of the command on the next line.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:32]: Since you've modified the "official" version of sleep, does it make sense to rename the file name to emphasize that it is a custom version for this demo? Maybe sleep-spire-demo.yaml or sleep-spire.yaml?

If istio-config.yaml was also modified, it could also be renamed.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:58]: It looks a little weird to see the output indented like this but I assume that's what the output really looks like.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:63]:
```suggestion
1.  Delete all deployments and configurations for the SPIRE Agent, Server, and namespace:
```
Just a minor suggestion. If you omit SPIRE (leave it as is) then agent and server can remain lowercase.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:3]: I'd prefer to have this link as the third paragraph rather than the first. My reason is that if a user is browsing through GitHub and wants to get a quick overview of what this repo does, it's more friendly to have that quick summary as the very first paragraph.

Even if you prefer not to move this paragraph, let's change the link to the page title of the other page instead of "here."
```suggestion
See [Istio CA Integration with SPIRE](https://istio.io/latest/docs/ops/integrations/spire) for further details about this integration .
```
Click [here](https://istio.io/latest/docs/releases/contribute/style-guide/#create-useful-links) for my rationale. 😉

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:8]: The Istio [doc guidelines](https://istio.io/latest/docs/releases/contribute/style-guide/#avoid-using-we) discourage using "we." I know many of the SPIRE docs do use "we," but it's probably best to follow Istio's guidelines.
```suggestion
Once SPIRE is deployed and integrated with Istio, this sample deploys a modified version of the [sleep](/samples/sleep/README.md) service and validates that its [identity](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-verifiable-identity-document-svid) was issued by SPIRE. 
```

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:15]: I can't remember if we talked about this. I just want to double check that `spire-quickstart.yaml` will be in the samples/security/spire directory along with this README and the other YAML files. I don't see it in GitHub but maybe I'm missing something.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:10]: A minor suggestion: "Usage" is fine but kind of nerdy. I'd prefer a more descriptive heading such as one of the following:

- Run the sample
- Run the sample test
- Deploy the sample
- Demonstrate how SPIRE issues an identity through Envoy's SDS API

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:45]: Please delete extra line here. (Picky, picky 🔍 )

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:47]: Or "using the"
```suggestion
5. Retrieve sleep's SVID identity document with the `istioctl proxy-config secret` command:
```
I know identity document is redundant since SVID means "SPIFFE Verifiable Identity Document". But if new users aren't familiar with SVID, those words can clarify it a little.

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:40]:
```suggestion
    Otherwise, manually inject the sidecars before applying:
```

File: samples/security/spire/README.md
Comment (sanderson042) [inline samples/security/spire/README.md:20]: Please reflow these two lines to make this one paragraph and change "until" to "before." (If you wanted to use "until" I think it would sound more correct like this: "Don't move to the next step until the deployment is completed." But I like the "before" version below.)
```suggestion
2. Ensure that the deployment is completed before moving to the next step. This can be verified by waiting on the `spire-agent` pod to become ready:
```

**Review by sanderson042 — COMMENTED** [review]
Hi @jufantozzi 👋 - I've added some more suggestions. But the current version of the doc is still quite good and really improved over the first version. 👍 It's pleasantly short and has the right amount of info for the context. For more info, users can see the other doc you've linked to.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:32]: I agree for the sleep service, I'm still not fully convinced for istio-config. My argument would be that, yes, for sleep, we have something that's really "close" to it (same directory/level) so some confusion could happen there, but there are no other istio-config's to get confused with... Also, the name istio-spire-config.yaml might be too big? IDK, any thoughts?

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:58]: indeed.

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:3]: I love to find depth in stuff that I never gave a second thought. Thanks for all the insights this gave me!

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:15]: They weren't initially, they should be there now 😄

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:10]: I totally agree with nerdy 😆

File: samples/security/spire/README.md
Comment (jufantozzi) [reply samples/security/spire/README.md:32]: Briefly discussed this with @alexandrealvino and now I'm convinced, lets go with all including spire in the name

File: samples/security/spire/README.md
Comment (mchurichi) [reply samples/security/spire/README.md:51]: How about this?
```suggestion
    $ istioctl pc secret $SLEEP_POD -o json \
      | jq -r '.dynamicActiveSecrets[0].secret.tlsCertificate.certificateChain.inlineBytes' \
      | base64 --decode > chain.pem
```

**Pull request closed without merging**
